bot/exts/moderation/clean.py

```diff
@@ -0,0 +1,589 @@
+import logging
+import re
+import time
+from collections import defaultdict
+from contextlib import suppress
+from datetime import datetime
+from itertools import islice
+from typing import Any, Callable, DefaultDict, Iterable, Literal, Optional, TYPE_CHECKING, Union
+
+from discord import Colour, Message, NotFound, TextChannel, User, errors
+from discord.ext.commands import Cog, Context, Converter, Greedy, group, has_any_role
+from discord.ext.commands.converter import TextChannelConverter
+from discord.ext.commands.errors import BadArgument
+
+from bot.bot import Bot
+from bot.constants import (
+    Channels, CleanMessages, Colours, Emojis, Event, Icons, MODERATION_ROLES
+)
+from bot.converters import Age, ISODateTime
+from bot.exts.moderation.modlog import ModLog
+from bot.utils.channel import is_mod_channel
+
+log = logging.getLogger(__name__)
+
+DEFAULT_TRAVERSE = 10
+MESSAGE_DELETE_DELAY = 5
+
+# Type alias for checks
+Predicate = Callable[[Message], bool]
+
+CleanLimit = Union[Message, Age, ISODateTime]
+
+
+class CleanChannels(Converter):
+    """A converter that turns the given string to a list of channels to clean, or the literal `*` for all channels."""
+
+    _channel_converter = TextChannelConverter()
+
+    async def convert(self, ctx: Context, argument: str) -> Union[Literal["*"], list[TextChannel]]:
+        """Converts a string to a list of channels to clean, or the literal `*` for all channels."""
+        if argument == "*":
+            return "*"
+        return [await self._channel_converter.convert(ctx, channel) for channel in argument.split()]
+
+
+class Regex(Converter):
+    """A converter that takes a string in the form `.+` and returns the contents of the inline code."""
+
+    async def convert(self, ctx: Context, argument: str) -> str:
+        """Strips the backticks from the string."""
+        return re.fullmatch(r"`(.+?)`", argument).group(1)
```

Comment:
We should handle invalid regex string here i.e. it is not in form of "\`.+\`".

Reply:
If it's not in that form, the match will fail and `group(1)` will raise an exception, causing the overall conversion to just fail.

Reply:
Scratch that, the exception seems to propagate when it's a required argument. Will fix it 👍